File: ebu-tt_to_srt.pl
#!/usr/bin/env perl

###############################################################################
# ebu-tt_to_srt.pl
#
# EBU-TT / TTML -> SRT converter
#
# Version 2.0
#
# Features:
#   - XML::LibXML based parser
#   - proper namespace handling
#   - UTF-8 input/output
#   - automatic output filename
#   - explicit input/output arguments
#   - automatic documentStartOfProgramme detection
#   - explicit --offset override
#   - clock-time timestamps
#   - media/offset timestamps
#   - frame-based timestamps
#   - configurable frame rate
#   - frameRate / frameRateMultiplier / subFrameRate detection
#   - <br/> -> SRT line break
#   - nested <span> handling
#   - XML whitespace handling
#   - empty subtitle detection
#   - invalid timestamp detection
#   - negative timestamps clamped to zero
#   - malformed subtitle handling
#   - optional strict mode
#   - optional verbose diagnostics
#
# Dependencies:
#   XML::LibXML
#
# Example:
#   ./ebu-tt_to_srt.pl input.xml
#
#   ./ebu-tt_to_srt.pl input.xml output.srt
#
#   ./ebu-tt_to_srt.pl --offset 10:00:00 input.xml output.srt
#
###############################################################################

use strict;
use warnings;
use utf8;

use Getopt::Long qw(GetOptions);
use File::Basename qw(fileparse);
use XML::LibXML;

use constant {
    VERSION => '2.0.0',
};

###############################################################################
# Configuration
###############################################################################

my %opt = (
    output       => undef,
    offset       => undef,
    fps          => undef,
    encoding     => 'UTF-8',
    strict       => 0,
    verbose      => 0,
    keep_empty   => 0,
    no_auto_offset => 0,
);

###############################################################################
# Command line
###############################################################################

sub usage {
    return <<"EOF";

ebu-tt_to_srt.pl v${\VERSION}

Convert EBU-TT / TTML subtitles to SRT.

Usage:

    ebu-tt_to_srt.pl [options] input.xml [output.srt]

Options:

    -o, --output FILE
        Output SRT filename.

    --offset TIME
        Explicit timestamp offset.

        Example:
            --offset 10:00:00
            --offset 10:00:00.000

    --fps RATE
        Explicit frame rate for frame-based timestamps.

        Examples:
            --fps 25
            --fps 25/1
            --fps 30000/1001

        If omitted, the value is taken from ttp:frameRate
        and ttp:frameRateMultiplier when available.

    --no-auto-offset
        Do not use ebuttm:documentStartOfProgramme automatically.

    --keep-empty
        Keep subtitles whose text content is empty.

    --strict
        Stop on malformed subtitle data instead of skipping it.

    -v, --verbose
        Print diagnostic information.

    -h, --help
        Show this help.

    --version
        Show program version.

Examples:

    ebu-tt_to_srt.pl subtitles.xml

    ebu-tt_to_srt.pl subtitles.xml subtitles.srt

    ebu-tt_to_srt.pl \\
        --offset 10:00:00 \\
        subtitles.xml subtitles.srt

    ebu-tt_to_srt.pl \\
        --fps 25 \\
        subtitles.xml subtitles.srt

EOF
}

GetOptions(
    'output|o=s'       => \$opt{output},
    'offset=s'         => \$opt{offset},
    'fps=s'            => \$opt{fps},
    'strict'           => \$opt{strict},
    'verbose|v'        => \$opt{verbose},
    'keep-empty'       => \$opt{keep_empty},
    'no-auto-offset'   => \$opt{no_auto_offset},
    'help|h'           => sub {
        print usage();
        exit 0;
    },
    'version'          => sub {
        print VERSION, "\n";
        exit 0;
    },
) or die usage();

my @args = @ARGV;

die usage() unless @args;

my $input = shift @args;

die "Unexpected argument(s): @args\n"
    if @args > 1;

if (!-f $input) {
    die "Input file does not exist: $input\n";
}

###############################################################################
# Output filename
###############################################################################

if (defined $opt{output}) {
    $opt{output} = $opt{output};
}
elsif (@args) {
    $opt{output} = $args[0];
}
else {
    my ($name) = fileparse($input, qr/\.[^.]*/);
    $opt{output} = $name . '.srt';
}

###############################################################################
# XML parser
###############################################################################

my $parser = XML::LibXML->new(
    recover      => 0,
    no_network   => 1,
    load_ext_dtd => 0,
    expand_entities => 0,
);

my $doc;

eval {
    $doc = $parser->parse_file($input);
};

if ($@ || !$doc) {
    my $error = $@ || 'unknown XML parsing error';
    chomp $error;

    die "Unable to parse XML file '$input': $error\n";
}

###############################################################################
# Namespace handling
###############################################################################

my $xpc = XML::LibXML::XPathContext->new($doc);

# TTML namespace
$xpc->registerNs(
    'tt',
    'http://www.w3.org/ns/ttml'
);

# TTML parameter namespace
$xpc->registerNs(
    'ttp',
    'http://www.w3.org/ns/ttml#parameter'
);

# EBU-TT metadata namespace
$xpc->registerNs(
    'ebuttm',
    'urn:ebu:tt:metadata'
);

###############################################################################
# Document parameters
###############################################################################

my $root = $doc->documentElement;

my $time_base = $root->getAttribute('timeBase');

$time_base = 'media'
    unless defined($time_base) && length($time_base);

my $frame_rate = detect_frame_rate(
    root => $root,
    xpc  => $xpc,
);

if (defined $opt{fps}) {
    $frame_rate = parse_frame_rate($opt{fps});
}

if ($opt{verbose}) {
    print STDERR "Input:       $input\n";
    print STDERR "Output:      $opt{output}\n";
    print STDERR "timeBase:    $time_base\n";
    print STDERR "frame rate:  $frame_rate\n";
}

###############################################################################
# Determine timestamp offset
###############################################################################

my $offset = 0;

if (defined $opt{offset}) {

    $offset = parse_timestamp(
        $opt{offset},
        frame_rate => $frame_rate,
        time_base  => $time_base,
    );

    verbose(
        "Using explicit offset: " .
        format_seconds($offset)
    );
}
elsif (!$opt{no_auto_offset}) {

    my $document_start =
        find_document_start_of_programme($xpc);

    if (defined $document_start) {

        eval {
            $offset = parse_timestamp(
                $document_start,
                frame_rate => $frame_rate,
                time_base  => $time_base,
            );
        };

        if ($@) {
            my $error = $@;
            chomp $error;

            warn
                "Warning: unable to parse " .
                "documentStartOfProgramme '$document_start': " .
                "$error\n";

            $offset = 0;
        }
        else {
            verbose(
                "Using documentStartOfProgramme: " .
                "$document_start (" .
                format_seconds($offset) .
                ")"
            );
        }
    }
}

###############################################################################
# Find subtitles
###############################################################################

my @paragraphs = $xpc->findnodes(
    '/tt:tt/tt:body//tt:p'
);

if (!@paragraphs) {
    die "No TTML subtitle paragraphs (<tt:p>) found.\n";
}

verbose("Found " . scalar(@paragraphs) . " subtitle(s)");

###############################################################################
# Open output
###############################################################################

open(
    my $out,
    '>:encoding(UTF-8)',
    $opt{output}
)
or die "Unable to open '$opt{output}' for writing: $!\n";

###############################################################################
# Convert subtitles
###############################################################################

my $srt_number = 1;
my $written    = 0;
my $skipped    = 0;

for my $p (@paragraphs) {

    my $begin = $p->getAttribute('begin');
    my $end   = $p->getAttribute('end');
    my $dur   = $p->getAttribute('dur');

    #
    # Some TTML documents may use begin + dur instead of begin + end.
    #
    if (
        defined($begin) &&
        length($begin) &&
        (!defined($end) || !length($end)) &&
        defined($dur) &&
        length($dur)
    ) {
        eval {

            my $begin_seconds = parse_timestamp(
                $begin,
                frame_rate => $frame_rate,
                time_base  => $time_base,
            );

            my $duration_seconds = parse_timestamp(
                $dur,
                frame_rate => $frame_rate,
                time_base  => $time_base,
                duration   => 1,
            );

            $end =
                format_seconds(
                    $begin_seconds + $duration_seconds
                );
        };

        if ($@) {
            handle_error(
                "Unable to calculate end time for subtitle",
                $p,
            );

            $skipped++;
            next;
        }
    }

    unless (
        defined($begin) && length($begin) &&
        defined($end)   && length($end)
    ) {
        handle_error(
            "Subtitle has no valid begin/end timing",
            $p,
        );

        $skipped++;
        next;
    }

    my ($begin_seconds, $end_seconds);

    eval {

        $begin_seconds = parse_timestamp(
            $begin,
            frame_rate => $frame_rate,
            time_base  => $time_base,
        );

        $end_seconds = parse_timestamp(
            $end,
            frame_rate => $frame_rate,
            time_base  => $time_base,
        );
    };

    if ($@) {
        handle_error(
            "Invalid subtitle timestamp: " . $@,
            $p,
        );

        $skipped++;
        next;
    }

    #
    # Apply programme offset.
    #
    $begin_seconds -= $offset;
    $end_seconds   -= $offset;

    #
    # SRT does not support negative timestamps.
    #
    $begin_seconds = 0
        if $begin_seconds < 0;

    $end_seconds = 0
        if $end_seconds < 0;

    #
    # Prevent malformed SRT intervals.
    #
    if ($end_seconds < $begin_seconds) {

        handle_error(
            sprintf(
                "Subtitle ends before it begins: %.3f > %.3f",
                $begin_seconds,
                $end_seconds,
            ),
            $p,
        );

        $skipped++;
        next;
    }

    #
    # Extract textual content.
    #
    my $text = extract_text($p);

    unless (length($text)) {

        unless ($opt{keep_empty}) {
            verbose(
                "Skipping empty subtitle"
            );

            $skipped++;
            next;
        }
    }

    #
    # Write SRT.
    #
    print {$out} $srt_number, "\n";

    print {$out}
        format_srt_time($begin_seconds),
        " --> ",
        format_srt_time($end_seconds),
        "\n";

    print {$out} $text, "\n\n";

    $srt_number++;
    $written++;
}

close($out)
or die "Unable to close '$opt{output}': $!\n";

###############################################################################
# Summary
###############################################################################

print
    "Written SRT: $opt{output}\n",
    "Subtitles:   $written\n";

print
    "Skipped:     $skipped\n"
    if $skipped;

exit 0;

###############################################################################
# Functions
###############################################################################

sub verbose {
    return unless $opt{verbose};

    print STDERR "[ebu-tt_to_srt] ", @_, "\n";
}

###############################################################################

sub handle_error {
    my ($message, $node) = @_;

    if ($opt{strict}) {
        die "$message\n";
    }

    warn "Warning: $message\n";
}

###############################################################################

sub detect_frame_rate {
    my (%args) = @_;

    my $root = $args{root};

    my $frame_rate =
        $root->getAttribute('frameRate');

    my $multiplier =
        $root->getAttribute('frameRateMultiplier');

    #
    # EBU-TT / TTML uses:
    #
    # frameRate
    # frameRateMultiplier
    #
    # Example:
    #
    # frameRate="30000"
    # frameRateMultiplier="1001 1000"
    #
    if (
        defined($frame_rate) &&
        length($frame_rate)
    ) {

        my $fps =
            parse_frame_rate($frame_rate);

        if (
            defined($multiplier) &&
            length($multiplier)
        ) {

            if (
                $multiplier =~
                /^\s*(\d+)\s+(\d+)\s*$/
            ) {

                my ($num, $den) =
                    ($1, $2);

                die
                    "Invalid frameRateMultiplier"
                    if $den == 0;

                $fps *= $num / $den;
            }
        }

        return $fps;
    }

    #
    # Common broadcast default.
    #
    return 25;
}

###############################################################################

sub parse_frame_rate {
    my ($value) = @_;

    if ($value =~ /^\s*(\d+(?:\.\d+)?)\s*$/) {
        my $fps = $1 + 0;

        die "Frame rate must be > 0\n"
            if $fps <= 0;

        return $fps;
    }

    if ($value =~ /^\s*(\d+)\s*\/\s*(\d+)\s*$/) {

        my ($num, $den) =
            ($1, $2);

        die "Frame rate denominator must not be zero\n"
            if $den == 0;

        return $num / $den;
    }

    die "Invalid frame rate: $value\n";
}

###############################################################################

sub find_document_start_of_programme {
    my ($xpc) = @_;

    #
    # EBU-TT metadata commonly appears as:
    #
    # <ebuttm:documentStartOfProgramme>
    #     10:00:00:00
    # </ebuttm:documentStartOfProgramme>
    #
    # We also tolerate an attribute representation.
    #

    my @nodes = $xpc->findnodes(
        '//*[local-name()="documentStartOfProgramme"]'
    );

    for my $node (@nodes) {

        #
        # Attribute form
        #
        my $value =
            $node->getAttribute('value');

        return $value
            if defined($value) && length($value);

        #
        # Text form
        #
        $value = $node->textContent;

        if (defined($value)) {

            $value =~ s/^\s+//;
            $value =~ s/\s+$//;

            return $value
                if length($value);
        }
    }

    #
    # Some documents use the metadata attribute directly.
    #
    my @attributes =
        $xpc->findnodes(
            '//@ebuttm:documentStartOfProgramme'
        );

    for my $attribute (@attributes) {

        my $value = $attribute->value;

        return $value
            if defined($value) && length($value);
    }

    return undef;
}

###############################################################################

sub parse_timestamp {
    my ($value, %args) = @_;

    my $fps =
        $args{frame_rate} || 25;

    my $time_base =
        $args{time_base} || 'media';

    die "Undefined timestamp\n"
        unless defined($value);

    $value =~ s/^\s+//;
    $value =~ s/\s+$//;

    #
    # Clock time:
    #
    # HH:MM:SS
    # HH:MM:SS.mmm
    # HH:MM:SS:FF
    #
    if (
        $value =~
        /^(\d+):(\d{2}):(\d{2})(?:\.(\d+))?$/
    ) {

        my ($hours, $minutes, $seconds, $fraction) =
            ($1, $2, $3, $4);

        die "Invalid minutes in timestamp '$value'\n"
            if $minutes > 59;

        die "Invalid seconds in timestamp '$value'\n"
            if $seconds > 59;

        my $fraction_seconds = 0;

        if (defined $fraction) {
            $fraction_seconds =
                ("0.$fraction") + 0;
        }

        return
            $hours * 3600 +
            $minutes * 60 +
            $seconds +
            $fraction_seconds;
    }

    #
    # Clock time with frames:
    #
    # HH:MM:SS:FF
    #
    if (
        $value =~
        /^(\d+):(\d{2}):(\d{2}):(\d+)$/
    ) {

        my ($hours, $minutes, $seconds, $frames) =
            ($1, $2, $3, $4);

        die "Invalid minutes in timestamp '$value'\n"
            if $minutes > 59;

        die "Invalid seconds in timestamp '$value'\n"
            if $seconds > 59;

        die "Invalid frame number '$frames'\n"
            if $frames >= $fps;

        return
            $hours * 3600 +
            $minutes * 60 +
            $seconds +
            ($frames / $fps);
    }

    #
    # Offset time expressions:
    #
    # 10h
    # 500ms
    # 10s
    # 25m
    # 100f
    #
    if ($value =~ /^(\d+(?:\.\d+)?)(h|m|s|ms|f)$/) {

        my ($number, $unit) =
            ($1, $2);

        return $number * 3600
            if $unit eq 'h';

        return $number * 60
            if $unit eq 'm';

        return $number
            if $unit eq 's';

        return $number / 1000
            if $unit eq 'ms';

        return $number / $fps
            if $unit eq 'f';
    }

    die
        "Unsupported timestamp expression '$value'\n";
}

###############################################################################

sub format_seconds {
    my ($seconds) = @_;

    return sprintf(
        "%.3f",
        $seconds,
    );
}

###############################################################################

sub format_srt_time {
    my ($seconds) = @_;

    $seconds = 0
        if $seconds < 0;

    #
    # SRT uses millisecond precision.
    #
    my $milliseconds =
        int($seconds * 1000 + 0.5);

    my $hours =
        int($milliseconds / 3_600_000);

    $milliseconds %= 3_600_000;

    my $minutes =
        int($milliseconds / 60_000);

    $milliseconds %= 60_000;

    my $secs =
        int($milliseconds / 1000);

    $milliseconds %= 1000;

    return sprintf(
        '%02d:%02d:%02d,%03d',
        $hours,
        $minutes,
        $secs,
        $milliseconds,
    );
}

###############################################################################

sub extract_text {
    my ($paragraph) = @_;

    my $text = '';

    #
    # Walk all direct/descendant children in document order.
    #
    # This is important because TTML allows:
    #
    # <p>
    #   Hello
    #   <span>world</span>
    #   <br/>
    #   <span>again</span>
    # </p>
    #
    for my $node ($paragraph->childNodes()) {

        $text .= extract_node_text($node);
    }

    #
    # Normalize CR/LF.
    #
    $text =~ s/\r\n/\n/g;
    $text =~ s/\r/\n/g;

    #
    # Remove trailing spaces from lines.
    #
    $text =~ s/[ \t]+\n/\n/g;

    #
    # Remove spaces introduced immediately after line breaks.
    #
    $text =~ s/\n[ \t]+/\n/g;

    #
    # Collapse excessive blank lines.
    #
    $text =~ s/\n{3,}/\n\n/g;

    #
    # Remove whitespace surrounding the complete subtitle.
    #
    $text =~ s/^\s+//;
    $text =~ s/\s+$//;

    return $text;
}

###############################################################################

sub extract_node_text {
    my ($node) = @_;

    #
    # Text node
    #
    if ($node->nodeType == XML_TEXT_NODE) {

        return $node->data;
    }

    #
    # Explicit TTML line break
    #
    if (
        $node->nodeType == XML_ELEMENT_NODE &&
        $node->localname eq 'br'
    ) {
        return "\n";
    }

    #
    # Other elements: recursively process children.
    #
    my $text = '';

    for my $child ($node->childNodes()) {
        $text .= extract_node_text($child);
    }

    return $text;
}

###############################################################################

File: ebu_tt_to_srt.py
#!/usr/bin/env python3
"""EBU-TT / TTML -> SRT converter, version 2.0.0."""

from __future__ import annotations

import argparse
import re
import sys
from dataclasses import dataclass
from decimal import Decimal, InvalidOperation, ROUND_HALF_UP
from pathlib import Path
from typing import Iterable, Optional

from lxml import etree

VERSION = "2.0.0"
TT_NS = "http://www.w3.org/ns/ttml"
TTP_NS = "http://www.w3.org/ns/ttml#parameter"

_CLOCK_RE = re.compile(r"^(\d+):(\d{2}):(\d{2})(?:\.(\d+))?$")
_CLOCK_FRAME_RE = re.compile(r"^(\d+):(\d{2}):(\d{2}):(\d+)$")
_OFFSET_RE = re.compile(r"^(\d+(?:\.\d+)?)(ms|h|m|s|f)$")
_FPS_RE = re.compile(r"^\s*(\d+(?:\.\d+)?)(?:\s*/\s*(\d+(?:\.\d+)?))?\s*$")


class EbuTTError(Exception):
    """Base converter error."""


class TimestampError(EbuTTError):
    """Invalid timestamp."""


@dataclass(frozen=True)
class Cue:
    number: int
    begin: Decimal
    end: Decimal
    text: str


@dataclass(frozen=True)
class ConversionResult:
    output: Path
    written: int
    skipped: int
    offset: Decimal
    frame_rate: Decimal
    time_base: str


def parse_frame_rate(value) -> Decimal:
    if isinstance(value, Decimal):
        fps = value
    else:
        m = _FPS_RE.match(str(value))
        if not m:
            raise EbuTTError(f"invalid frame rate: {value!r}")
        try:
            fps = Decimal(m.group(1)) / Decimal(m.group(2) or "1")
        except (InvalidOperation, ZeroDivisionError) as exc:
            raise EbuTTError(f"invalid frame rate: {value!r}") from exc
    if fps <= 0:
        raise EbuTTError("frame rate must be greater than zero")
    return fps


def detect_frame_rate(root: etree._Element) -> Decimal:
    raw = root.get(f"{{{TTP_NS}}}frameRate")
    if not raw:
        return Decimal("25")
    fps = parse_frame_rate(raw)
    mult = root.get(f"{{{TTP_NS}}}frameRateMultiplier")
    if mult:
        parts = mult.split()
        if len(parts) != 2:
            raise EbuTTError(f"invalid frameRateMultiplier: {mult!r}")
        try:
            fps *= Decimal(parts[0]) / Decimal(parts[1])
        except (InvalidOperation, ZeroDivisionError) as exc:
            raise EbuTTError(f"invalid frameRateMultiplier: {mult!r}") from exc
    return fps


def parse_timestamp(value: str, *, frame_rate: Decimal,
                    time_base: str = "media") -> Decimal:
    if value is None:
        raise TimestampError("timestamp is missing")
    value = value.strip()
    if not value:
        raise TimestampError("timestamp is empty")

    m = _CLOCK_RE.match(value)
    if m:
        h, mi, s = map(int, m.group(1, 2, 3))
        if mi > 59 or s > 59:
            raise TimestampError(f"invalid timestamp: {value!r}")
        frac = Decimal("0." + m.group(4)) if m.group(4) else Decimal(0)
        return Decimal(h * 3600 + mi * 60 + s) + frac

    m = _CLOCK_FRAME_RE.match(value)
    if m:
        h, mi, s, frames = map(int, m.groups())
        if mi > 59 or s > 59:
            raise TimestampError(f"invalid timestamp: {value!r}")
        if Decimal(frames) >= frame_rate:
            raise TimestampError(f"frame {frames} outside frame rate {frame_rate}")
        return Decimal(h * 3600 + mi * 60 + s) + Decimal(frames) / frame_rate

    m = _OFFSET_RE.match(value)
    if m:
        n, unit = Decimal(m.group(1)), m.group(2)
        return {
            "h": n * 3600, "m": n * 60, "s": n,
            "ms": n / 1000, "f": n / frame_rate,
        }[unit]

    # Real-world media feeds sometimes use plain seconds.
    if time_base == "media" and re.fullmatch(r"\d+(?:\.\d+)?", value):
        return Decimal(value)

    raise TimestampError(f"unsupported timestamp: {value!r}")


def format_srt_time(seconds: Decimal) -> str:
    seconds = max(Decimal(0), seconds)
    ms = int((seconds * 1000).quantize(Decimal(1), rounding=ROUND_HALF_UP))
    h, rem = divmod(ms, 3_600_000)
    m, rem = divmod(rem, 60_000)
    s, ms = divmod(rem, 1000)
    return f"{h:02d}:{m:02d}:{s:02d},{ms:03d}"


def _local_name(node: etree._Element) -> str:
    return etree.QName(node).localname


def _document_start(root: etree._Element) -> Optional[str]:
    names = {"documentStartOfProgramme", "documentStartOfProgram"}
    for node in root.iter():
        if _local_name(node) in names:
            value = (node.text or "").strip()
            if value:
                return value
            for attr in ("value", "time"):
                value = (node.get(attr) or "").strip()
                if value:
                    return value
    for node in root.iter():
        for attr, value in node.attrib.items():
            if etree.QName(attr).localname in names and value.strip():
                return value.strip()
    return None


def _timing_value(value: str, fps: Decimal, time_base: str) -> Decimal:
    return parse_timestamp(value, frame_rate=fps, time_base=time_base)


def _resolve_timing(p: etree._Element, fps: Decimal,
                    time_base: str) -> tuple[Decimal, Decimal]:
    # EBU-TT permits timing on body/div/p and child expressions can be
    # relative to the nearest timed ancestor. Accumulate ancestor begins.
    ancestors = list(p.iterancestors())
    ancestors.reverse()
    chain = ancestors + [p]

    current_begin = Decimal(0)
    parent_end: Optional[Decimal] = None

    for node in chain:
        b = node.get("begin")
        e = node.get("end")
        d = node.get("dur")

        node_start = current_begin
        if b:
            current_begin += _timing_value(b, fps, time_base)

        if e:
            parent_end = node_start + _timing_value(e, fps, time_base)
        elif d:
            parent_end = current_begin + _timing_value(d, fps, time_base)

    p_begin = current_begin
    b = p.get("begin")
    p_parent_begin = Decimal(0)
    for node in chain[:-1]:
        if node.get("begin"):
            p_parent_begin += _timing_value(node.get("begin"), fps, time_base)

    if p.get("end"):
        p_end = p_parent_begin + _timing_value(p.get("end"), fps, time_base)
    elif p.get("dur"):
        p_end = p_begin + _timing_value(p.get("dur"), fps, time_base)
    elif parent_end is not None:
        p_end = parent_end
    else:
        raise TimestampError("subtitle has neither end nor dur")

    if p_end < p_begin:
        raise TimestampError("subtitle end precedes begin")
    return p_begin, p_end


def _text(node: etree._Element) -> str:
    parts = []
    if node.text:
        parts.append(node.text)
    for child in node:
        if _local_name(child) == "br":
            parts.append("\n")
        else:
            parts.append(_text(child))
        if child.tail:
            parts.append(child.tail)
    return "".join(parts)


def extract_text(p: etree._Element) -> str:
    text = _text(p)
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    text = re.sub(r"[ \t]+\n", "\n", text)
    text = re.sub(r"\n[ \t]+", "\n", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip()


def _parse_xml(path: Path) -> etree._ElementTree:
    parser = etree.XMLParser(
        resolve_entities=False,
        no_network=True,
        load_dtd=False,
        recover=False,
    )
    try:
        return etree.parse(str(path), parser)
    except (OSError, etree.XMLSyntaxError) as exc:
        raise EbuTTError(f"unable to parse XML '{path}': {exc}") from exc


def convert(input_path, output_path=None, *, offset=None, fps=None,
            no_auto_offset=False, strict=False, verbose=False,
            keep_empty=False) -> ConversionResult:
    source = Path(input_path)
    if not source.is_file():
        raise EbuTTError(f"input file does not exist: {source}")

    destination = Path(output_path) if output_path else source.with_suffix(".srt")
    tree = _parse_xml(source)
    root = tree.getroot()

    time_base = root.get(f"{{{TTP_NS}}}timeBase", "media").lower()
    if time_base not in {"media", "clock", "smpte"}:
        if strict:
            raise EbuTTError(f"unsupported timeBase: {time_base!r}")
        if verbose:
            print(f"Warning: unsupported timeBase {time_base!r}; using media",
                  file=sys.stderr)
        time_base = "media"

    frame_rate = parse_frame_rate(fps) if fps is not None else detect_frame_rate(root)

    offset_seconds = Decimal(0)
    if offset is not None:
        offset_seconds = (offset if isinstance(offset, Decimal)
                          else parse_timestamp(str(offset), frame_rate=frame_rate,
                                               time_base=time_base))
    elif not no_auto_offset:
        start = _document_start(root)
        if start:
            try:
                offset_seconds = parse_timestamp(start, frame_rate=frame_rate,
                                                 time_base=time_base)
            except TimestampError as exc:
                if strict:
                    raise
                if verbose:
                    print(f"Warning: invalid documentStartOfProgramme: {exc}",
                          file=sys.stderr)

    paragraphs = root.xpath(
        ".//*[local-name()='p' and namespace-uri()=$ns]", ns=TT_NS
    )
    if not paragraphs:
        raise EbuTTError("no TTML <p> subtitle elements found")

    cues = []
    skipped = 0

    for p in paragraphs:
        try:
            begin, end = _resolve_timing(p, frame_rate, time_base)
            text = extract_text(p)
            if not text and not keep_empty:
                skipped += 1
                continue

            begin -= offset_seconds
            end -= offset_seconds
            if end < 0:
                skipped += 1
                continue
            begin = max(Decimal(0), begin)
            end = max(Decimal(0), end)
            if end < begin:
                raise TimestampError("timestamp invalid after offset")

            cues.append(Cue(len(cues) + 1, begin, end, text))
        except (TimestampError, EbuTTError) as exc:
            if strict:
                raise
            skipped += 1
            print(f"Warning: skipping subtitle: {exc}", file=sys.stderr)

    destination.parent.mkdir(parents=True, exist_ok=True)
    with destination.open("w", encoding="utf-8", newline="\n") as fh:
        for cue in cues:
            fh.write(f"{cue.number}\n")
            fh.write(f"{format_srt_time(cue.begin)} --> "
                     f"{format_srt_time(cue.end)}\n")
            fh.write(cue.text + "\n\n")

    if verbose:
        print(f"[ebu-tt_to_srt] written={len(cues)} skipped={skipped}",
              file=sys.stderr)

    return ConversionResult(destination, len(cues), skipped,
                            offset_seconds, frame_rate, time_base)


def build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(description="Convert EBU-TT/TTML to SRT.")
    p.add_argument("input", type=Path)
    p.add_argument("output", type=Path, nargs="?")
    p.add_argument("-o", "--output", dest="output_option", type=Path)
    p.add_argument("--offset")
    p.add_argument("--fps")
    p.add_argument("--no-auto-offset", action="store_true")
    p.add_argument("--keep-empty", action="store_true")
    p.add_argument("--strict", action="store_true")
    p.add_argument("-v", "--verbose", action="store_true")
    p.add_argument("--version", action="version", version=VERSION)
    return p


def main(argv: Optional[Iterable[str]] = None) -> int:
    args = build_parser().parse_args(argv)
    if args.output and args.output_option:
        print("Error: specify output positionally or with --output, not both",
              file=sys.stderr)
        return 2
    output = args.output_option or args.output
    try:
        result = convert(
            args.input, output,
            offset=args.offset, fps=args.fps,
            no_auto_offset=args.no_auto_offset,
            strict=args.strict, verbose=args.verbose,
            keep_empty=args.keep_empty,
        )
    except (EbuTTError, OSError, ValueError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    print(f"Written {result.written} subtitle(s) to {result.output}"
          + (f"; skipped {result.skipped}" if result.skipped else ""))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
